File: dsa-problems/Sorting/solutions/java/SRT-007-search-rotated-duplicates-parity.java
import java.util.*;

class Solution {
    public int countEvenIndices(int[] arr, int x) {
        int n = arr.length;
        int pivot = findPivot(arr, 0, n - 1);
        
        int count = 0;
        // Search in left part [0, pivot-1]
        if (pivot > 0) {
            int[] range = searchRange(arr, 0, pivot - 1, x);
            if (range[0] != -1) {
                count += countEvens(range[0], range[1]);
            }
        }
        
        // Search in right part [pivot, n-1]
        int[] range = searchRange(arr, pivot, n - 1, x);
        if (range[0] != -1) {
            count += countEvens(range[0], range[1]);
        }
        
        return count;
    }
    
    private int findPivot(int[] arr, int low, int high) {
        // Returns index of minimum element
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] > arr[high]) {
                low = mid + 1;
            } else if (arr[mid] < arr[high]) {
                high = mid;
            } else {
                // Duplicates: worst case O(N)
                high--;
            }
        }
        return low;
    }
    
    private int[] searchRange(int[] arr, int low, int high, int target) {
        int start = -1, end = -1;
        
        // Find first
        int l = low, r = high;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (arr[mid] >= target) {
                if (arr[mid] == target) start = mid;
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        
        if (start == -1) return new int[]{-1, -1};
        
        // Find last
        l = low; r = high;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (arr[mid] <= target) {
                if (arr[mid] == target) end = mid;
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        
        return new int[]{start, end};
    }
    
    private int countEvens(int L, int R) {
        if (L > R) return 0;
        int len = R - L + 1;
        if (len % 2 == 0) return len / 2;
        return (L % 2 == 0) ? (len + 1) / 2 : (len - 1) / 2;
    }
}

File: dsa-problems/Sorting/solutions/python/SRT-007-search-rotated-duplicates-parity.py
def count_even_indices(arr: list[int], x: int) -> int:
    n = len(arr)
    
    def find_pivot(low, high):
        while low < high:
            mid = (low + high) // 2
            if arr[mid] > arr[high]:
                low = mid + 1
            elif arr[mid] < arr[high]:
                high = mid
            else:
                high -= 1
        return low
        
    def search_range(low, high, target):
        start, end = -1, -1

        # Find leftmost occurrence
        l, r = low, high
        while l <= r:
            mid = (l + r) // 2
            if arr[mid] == target:
                start = mid
                r = mid - 1
            elif arr[mid] < target:
                l = mid + 1
            else:
                r = mid - 1

        if start == -1:
            return -1, -1

        # Find rightmost occurrence
        l, r = low, high
        while l <= r:
            mid = (l + r) // 2
            if arr[mid] == target:
                end = mid
                l = mid + 1
            elif arr[mid] < target:
                l = mid + 1
            else:
                r = mid - 1

        return start, end
        
    def count_evens(L, R):
        if L > R: return 0
        length = R - L + 1
        if length % 2 == 0:
            return length // 2
        return (length + 1) // 2 if L % 2 == 0 else (length - 1) // 2

    pivot = find_pivot(0, n - 1)
    count = 0
    
    if pivot > 0:
        s, e = search_range(0, pivot - 1, x)
        if s != -1:
            count += count_evens(s, e)
            
    s, e = search_range(pivot, n - 1, x)
    if s != -1:
        count += count_evens(s, e)
        
    return count

def main():
    n, x = map(int, input().split())
    arr = list(map(int, input().split()))
    result = count_even_indices(arr, x)
    print(result)

if __name__ == "__main__":
    main()
